subsys/bluetooth/host/id.c

```diff
@@ -199,8 +200,13 @@ static void le_rpa_invalidate(void)
 		atomic_clear_bit(bt_dev.flags, BT_DEV_RPA_VALID);
 	}
 
+	/* Invalidate RPA of advertising sets that are disabled.
+	 *
+	 * Enabled advertising set have to call rpa_expired callback first to
+	 * determine if the RPA of the advertising set should be rotated.
+	 */
```

Comment:
Please consider my comment:

https://github.com/zephyrproject-rtos/zephyr/issues/51247#issuecomment-1283705406

Reply:
@Thalley, @cvinayak and other stakeholders/reviewers of the ```rpa_expired``` callback feature. I suggested the potential modification in the linked comment above. We discussed it briefly with @joerchan. Please speak up if you have any counter-arguments/feedback regarding the conditions in which this callback is called.

Reply:
I am not sure I follow the issue mentioned in that comment well enough to provide any additional comments at this point

Reply:
@jhedberg @jori-nordic @alwa-nordic @cvinayak we need input to resolve this.

Reply:
> Would it be useful for the application to get a new_rpa event? Right now rpa_expired is suggested as the event the application should use to synchronize with RPA changes. But that does not quite fit with the stopped-advertising-set case (unless we do assign new RPAs to stopped advertising sets), since a rotation would also happen on adv-start. Hopefully most applications are fine with just the 'expired' event.

I think the new_rpa event is unnecessary if the API user is familiar with Zephyr Bluetooth stack mechanics (we might improve the documentation for this reason). If the advertising is ongoing, the ```rpa_expired``` is also a signal that the new RPA will be used after the callback execution. If the advertising is stopped and the advertising set is not deleted, the ```rpa_expired``` callback should also notify the user about the RPA expiry. The API user should be aware at this point that the Bluetooth stack uses the new RPA upon the advertising restart.

Furthermore, the API user has full control over the RPA for its advertising set regardless of the advertising state (running / disabled). They can suspend the RPA rotation with the ```rpa_expired``` callback for an indefinite time while switching between advertising states. They can also enforce the rotation asynchronously with the ```bt_le_ext_adv_oob_get_local``` API. They have also the possibility to reject the RPA rotation in case of the asynchronous ```bt_le_ext_adv_oob_get_local``` API call, triggered by the user of some other advertising set (mitigates dependency on other advertising sets).

Reply:
Summary from @kapi-no:
The `rpa_expired` callback should be called every time when the `BT_ADV_RPA_VALID` flag is cleared for consistency (clearing causes RPA regeneration).